### PR TITLE
adds poisson NLL loss

### LINKED ISSUE #1774: Poisson NLL loss
There is no log-likelihood loss with underlying target's Poisson distribution. Would be great to add. I can do it.

### PR DESCRIPTION
**Pull request by vlasenkov**
Resolves #1774.

### CODE REVIEW

**Comment by ezyang** [comment]
Needs a test!

**Comment by vlasenkov** [comment]
tests fail when attempting to find `PoissonNLLLoss` in `torch.legacy.nn`. So, I need to add Poisson criterion there. Could you clarify why this legacy modules are separated from `torch.nn`?

**Comment by vlasenkov** [comment]
@ezyang, @fmassa,  some remarks?

File: torch/nn/functional.py [hunk 1]
```diff
@@ -547,6 +547,22 @@ def nll_loss(input, target, weight=None, size_average=True):
...
+def poisson_nll_loss(log_input, target, size_average=True):
+    r"""Poisson negative log likelihood loss.
+
+    Drops :math:`log(target!)` term. See :class:`~torch.nn.PoissonNLLLoss` for details.
+
+    Args:
+        log_input: :math:`log(input)`
+        target: :math:`target \sim Pois(input)`
+    """
+    loss = torch.exp(log_input) - target * log_input
```
Comment (fmassa) [inline torch/nn/functional.py]: Is there a reason why you expect `log(input)` instead of `input` as the input? Couldn't you do instead
```python
loss = input - target * torch.log(input)
```
?
Also, what's the relationship with `nll_loss`? Do you have a paper that proposes this loss?
Because if it is somewhat similar to `nll_loss`, you'd need to consider that the loss only takes into account the value of the input at a specified target index.

Comment (vlasenkov) [inline torch/nn/functional.py]: The loss accepts log input to avoid cases of application log to non-positive numbers. Examples: [Article](http://www.mathstat.dal.ca/~hgu/Neural%20Comput%20&%20Applic.pdf)

Comment (fmassa) [inline torch/nn/functional.py]: I see.
This loss seems a bit too niche, but as it's implemented in TF then maybe it would make sense to have it.
Maybe it would be good then to also add the stirling's approximation for the loss as an argument? It seems that we can easily implement it following [TF implementation](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/python/ops/nn_impl.py#L88-L96)?

Comment (vlasenkov) [inline torch/nn/functional.py]: I could do it, but this seems to have very restricted usage. The approximation requires evaluation of `log(target)`, but `target` for poissonian processes often happens to be zero. 
By the way probably we should add two losses? One is ordinary `PoissonNLLLoss` (without `log(input)`) and the other is `LogPoissonNLLLoss` which is current implementation

**Comment by vlasenkov** [comment]
@fmassa, PR is ready for review.
It's usual for target to be 2 or 3, and when summing loss for a large sample the error of approximation grows, so I wanted to add optional ability to compute exact `log(target!)` for small target values. This could be implemented as:
```
mask = target <= 10
loss[mask] += target[mask].apply_(lambda x: math.log(math.factorial(x)))
loss[~mask] += ...  # Stirling approximation
```
But `autograd.Variable` has no `apply` member. Are there some workarounds?

**Comment by vlasenkov** [comment]
@apaszke, could you comment on what is written above?

**Comment by fmassa** [comment]
@vlasenkov using an `apply_` function would be extremely slow on the GPU, so I wouldn't really advise on that. I'll have a look at your latest changes.
On a quick look, I think it looks ok. But I wonder if you don't have problems with autograd in `loss[mask] += ...`  which complains that you can't use an in-place operation there (I had sometimes to do things like `x[1:10] = x[1:10] + y` instead of `x[1:10] += y` for it to work).

**Comment by vlasenkov** [comment]
`loss[mask] += ...` works without problem. I've tested on OS X (cpu) and Ubuntu (cpu & gpu)

**Comment by vlasenkov** [comment]
ping @fmassa

**Comment by vlasenkov** [comment]
Can anyone merge this loss? Does it still need any fixes?

**Comment by fmassa** [comment]
I think this looks good, thanks!
@vlasenkov could you add link to the docs in https://github.com/pytorch/pytorch/blob/master/docs/source/nn.rst? If not they won't show in the website.

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks @vlasenkov !
